src/flinkStatementResultsManager.ts

```diff
@@ -222,19 +227,56 @@ export class FlinkStatementResultsManager {
     }
   }
 
+  /**
+   * Filters results based on the current search query and visible columns.
+   * If visibleColumns is undefined, searches through all columns.
+   * If visibleColumns is defined, only searches through the specified columns.
+   *
+   * @param results - Array of result rows to filter
+   * @param visibleColumns - Optional array of column names to search through. If undefined, searches all columns.
+   * @returns Filtered array of results that match the search query in the specified columns
+   */
+  private filterResultsBySearch(results: any[], visibleColumns: string[] | undefined): any[] {
+    const searchQuery = this._searchQuery();
+    if (searchQuery === null) {
+      return results;
+    }
+
+    const searchLower = searchQuery.toLowerCase();
+    return results.filter((row) =>
+      Object.entries(row)
+        .filter(([key]) => visibleColumns === undefined || visibleColumns.includes(key))
+        .some(([_, value]) => value !== null && String(value).toLowerCase().includes(searchLower)),
+    );
+  }
+
   handleMessage(type: MessageType, body: any): any {
     switch (type) {
       case "GetResults": {
         const offset = body.page * body.pageSize;
         const limit = body.pageSize;
         const paginatedResults = this.getResultsArray().slice(offset, offset + limit);
+        const filteredResults = this.filterResultsBySearch(paginatedResults, body.visibleColumns);
+
         return {
-          results: paginatedResults,
+          results: filteredResults,
         };
       }
       case "GetResultsCount": {
-        const count = this._results().size;
-        return { total: count, filter: null };
+        const results = this.getResultsArray();
+        const filteredResults = this.filterResultsBySearch(results, body.visibleColumns);
```

Comment:
Does this imply that whenever the user does a search, the backend essentially performs the filtering 2x, 'cause there's always going to be a call to request the results count also?

Perhaps having explicit data members for the filtered count + filtered rows, recalculated whenever setting the search string happens ...